src/lib.rs

```diff
@@ -434,76 +385,479 @@ impl<const A: bool> FromStr for ParsedIri<String, A> {
             positions: iri.positions,
         }
     }
 }
 
-impl<const A: bool> From<ParsedIri<Box<str>, A>> for ParsedIri<String, A> {
+impl From<IriRef<Box<str>>> for IriRef<String> {
     #[inline]
-    fn from(iri: ParsedIri<Box<str>, A>) -> Self {
+    fn from(iri: IriRef<Box<str>>) -> Self {
         Self {
             iri: iri.iri.into(),
             positions: iri.positions,
         }
     }
 }
 
-impl<'a, const A: bool> From<ParsedIri<&'a str, A>> for ParsedIri<Cow<'a, str>, A> {
+impl<'a> From<IriRef<&'a str>> for IriRef<Cow<'a, str>> {
     #[inline]
-    fn from(iri: ParsedIri<&'a str, A>) -> Self {
+    fn from(iri: IriRef<&'a str>) -> Self {
         Self {
             iri: iri.iri.into(),
             positions: iri.positions,
         }
     }
 }
 
-impl<'a, const A: bool> From<ParsedIri<String, A>> for ParsedIri<Cow<'a, str>, A> {
+impl<'a> From<IriRef<String>> for IriRef<Cow<'a, str>> {
     #[inline]
-    fn from(iri: ParsedIri<String, A>) -> Self {
+    fn from(iri: IriRef<String>) -> Self {
         Self {
             iri: iri.iri.into(),
             positions: iri.positions,
         }
     }
 }
 
-impl<'a, const A: bool> From<&'a ParsedIri<String, A>> for ParsedIri<&'a str, A> {
+impl<'a> From<&'a IriRef<String>> for IriRef<&'a str> {
     #[inline]
-    fn from(iri: &'a ParsedIri<String, A>) -> Self {
+    fn from(iri: &'a IriRef<String>) -> Self {
         Self {
             iri: &iri.iri,
             positions: iri.positions,
         }
     }
 }
 
-impl<'a, const A: bool> From<&'a ParsedIri<Cow<'a, str>, A>> for ParsedIri<&'a str, A> {
+impl<'a> From<&'a IriRef<Cow<'a, str>>> for IriRef<&'a str> {
     #[inline]
-    fn from(iri: &'a ParsedIri<Cow<'a, str>, A>) -> Self {
+    fn from(iri: &'a IriRef<Cow<'a, str>>) -> Self {
         Self {
             iri: &iri.iri,
             positions: iri.positions,
         }
     }
 }
 
+/// A [RFC 3987](https://www.ietf.org/rfc/rfc3987) IRI.
+///
+/// Instances of this type are guaranteed to be absolute,
+/// unlike [`IriRef`].
+///
+/// ```
+/// use oxiri::Iri;
+///
+/// // Parse and validate base IRI
+/// let base_iri = Iri::parse("http://foo.com/bar/baz").unwrap();
+///
+/// // Validate and resolve relative IRI
+/// let iri = base_iri.resolve("bat#foo").unwrap();
+/// assert_eq!(iri.into_inner(), "http://foo.com/bar/bat#foo");
+///
+/// // Iri::parse will err on relative IRIs.
+/// assert!(Iri::parse("../bar/baz").is_err())
+/// ```
+#[derive(Clone, Copy)]
+pub struct Iri<T>(IriRef<T>);
+
+impl<T: Deref<Target = str>> Iri<T> {
+    /// Parses and validates the IRI following the grammar from [RFC 3987](https://www.ietf.org/rfc/rfc3987).
+    ///
+    /// This operation keeps internally the `iri` parameter and does not allocate.
+    ///
+    /// ```
+    /// use oxiri::Iri;
+    ///
+    /// Iri::parse("http://foo.com/bar/baz").unwrap();
+    /// ```
+    pub fn parse(iri: T) -> Result<Self, IriParseError> {
+        IriRef::parse(iri)?.try_into()
```

Comment:
this will be less efficient than the current code: given a very long relative IRI, this will parse it entirely, before checking whether it is relative.
By contrast, the current code stop immediately when it fails to parse a scheme, when the parse mode is `Absolute`

Reply:
That's a good point. But is it a case we care about optimizing a lot? I am not sure it's meaningful to bother about the performance of unhappy paths.

Reply:
fair enough